src/text-fragment-utils.js

```diff
@@ -192,33 +208,124 @@ const processTextFragmentDirective = (textFragment) => {
     range.setEnd(endNode, endOffset);
     try {
       range.surroundContents(mark);
+      return [mark];
     } catch {
-      // Text to highlight does not contain entire DOM nodes.
-      // Need to extend the highlighted selection to entire nodes (Ex: entire links).
+      // Text to highlight does not contain entire DOM elements.
+      // Need to create multiple <mark> elements to highlight the entire selection.
       const commonAncestor = range.commonAncestorContainer;
-      while (startNode.parentNode !== commonAncestor) {
-        startNode = startNode.parentNode;
-        range.setStartBefore(startNode);
-      }
-      while (endNode.parentNode !== commonAncestor) {
-        endNode = endNode.parentNode;
-        range.setEndAfter(endNode);
+      const startMarks = [];
+      let firstNode = startNode;
+      startMarks.push(
+        createMarkFor(firstNode, {
+          start: { node: firstNode, offset: startOffset },
+        }),
+      );
+      firstNode = startMarks[0];
+      while (firstNode.parentNode !== commonAncestor) {
+        const parent = firstNode.parentNode;
+        let nodesToHighlight = Array.from(parent.childNodes);
+        nodesToHighlight = nodesToHighlight.slice(
+          nodesToHighlight.indexOf(firstNode) + 1,
+        );
+        startMarks.concat(highlightNodes(nodesToHighlight));
+        firstNode = parent;
       }
-      try {
-        range.surroundContents(mark);
-      } catch {
-        // Text highlight still didn't work.
-        return;
+      const endMarks = [];
+      let lastNode = endNode;
+      endMarks.push(
+        createMarkFor(lastNode, { end: { node: lastNode, offset: endOffset } }),
+      );
+      lastNode = endMarks[0];
+      while (lastNode.parentNode !== commonAncestor) {
+        const parent = lastNode.parentNode;
+        let nodesToHighlight = Array.from(parent.childNodes);
+        nodesToHighlight = nodesToHighlight.slice(
+          0,
+          nodesToHighlight.indexOf(lastNode),
+        );
+        endMarks.concat(highlightNodes(nodesToHighlight));
+        lastNode = parent;
       }
+      let nodesInBetween = Array.from(commonAncestor.childNodes);
+      nodesInBetween = nodesInBetween.slice(
+        nodesInBetween.indexOf(firstNode) + 1,
+        nodesInBetween.indexOf(lastNode),
+      );
+      return [
+        ...startMarks.filter(Boolean),
+        ...highlightNodes(nodesInBetween),
+        ...endMarks.filter(Boolean),
+      ];
     }
   }
   if (prefixNodes.length) {
     // ToDo
   }
-  if (mark.parentElement) {
+};
+
+/**
+ * Highlights the provided nodes entirely. Creates as few `<mark>` elements as possible.
+ * @param {Node[]} nodes
+ * @return {Element[]} mark elements created.
+ */
+const highlightNodes = (nodes) => {
+  if (!nodes || !nodes.length) return [];
+
+  const createdMarks = [];
+  let firstNode = nodes[0];
```

Comment:
What are `firstNode` and `lastNode` tracking? It's a bit confusing that they change, and frequently equal each other.